<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/images/SN_web_lightmode.png" width="300" alt="cognitiveclass.ai logo">
</center>

# Working with a real world data-set using SQL and Python

Estaimted time needed: **30** minutes

## Objectives

After complting this lab you will be able to:

-   Understand the dataset for Chicago Public School level performance
-   Store the dataset in an Db2 database on IBM Cloud instance
-   Retrieve metadata about tables and columns and query data from mixed case columns
-   Solve example problems to practice your SQL skills including using built-in database functions


## Chicago Public Schools - Progress Report Cards (2011-2012)

The city of Chicago released a dataset showing all school level performance data used to create School Report Cards for the 2011-2012 school year. The dataset is available from the Chicago Data Portal: [https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t](https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork22-2022-01-01&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

This dataset includes a large number of metrics. Start by familiarizing yourself with the types of metrics in the database: [https://data.cityofchicago.org/api/assets/AAD41A13-BE8A-4E67-B1F5-86E711E09D5F?download=true](https://data.cityofchicago.org/api/assets/AAD41A13-BE8A-4E67-B1F5-86E711E09D5F?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork22-2022-01-01&download=true&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

**NOTE**: 

Do not download the dataset directly from City of Chicago portal. Instead download a static copy which is a more database friendly version from this <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv">link</a>.

**NOTE**:

For the learners who are encountering issues with loading from .csv in DB2 on Firefox, you can download the .txt files and load the data with those: <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.txt">link</a>.

Now review some of its contents.


### Store the dataset in a Table

In many cases the dataset to be analyzed is available as a .CSV (comma separated values) file, perhaps on the internet. To analyze the data using SQL, it first needs to be stored in the database.

While it is easier to read the dataset into a Pandas dataframe and then PERSIST it into the database as we saw in the previous lab, it results in mapping to default datatypes which may not be optimal for SQL querying. For example a long textual field may map to a CLOB instead of a VARCHAR. 

Therefore, **it is highly recommended to manually load the table using the database console LOAD tool, as indicated in Week 2 Lab 1 Part II**. The only difference with that lab is that in Step 5 of the instructions you will need to click on create "(+) New Table" and specify the name of the table you want to create and then click "Next". 

##### Now open the Db2 console, open the LOAD tool, Select / Drag the .CSV file for the CHICAGO PUBLIC SCHOOLS dataset and load the dataset into a new table called **SCHOOLS**.

<a href="https://cognitiveclass.ai/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork22-2022-01-01"><img src="https://ibm.box.com/shared/static/uc4xjh1uxcc78ks1i18v668simioz4es.jpg"></a>


### Connect to the database

Let us now load the ipython-sql  extension and establish a connection with the database

The following modules are pre-installed in the Skills Network Labs environment. However if you run this notebook commands in a different Jupyter environment (e.g. Watson Studio or Ananconda) you may need to install these libraries by removing the `#` sign before `!pip` in the code cell below.


In [1]:
# These libraries are pre-installed in SN Labs. If running in another environment please uncomment lines below to install them:
# !pip install --force-reinstall ibm_db==3.1.0 ibm_db_sa==0.3.3
# Ensure we don't load_ext with sqlalchemy>=1.4 (incompadible)
# !pip uninstall sqlalchemy==1.4 -y && pip install sqlalchemy==1.3.24
# !pip install ipython-sql

In [2]:
%load_ext sql

In [3]:
# Enter the connection string for your Db2 on Cloud database instance below
# %sql ibm_db_sa://my-username:my-password@my-hostname:my-port/my-db-name?security=SSL
%sql ibm_db_sa://kkj13779:QCtjszJGs5wXfH6M@824dfd4d-99de-440d-9991-629c01b3832d.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:30119/bludb?security=SSL

'Connected: kkj13779@bludb'

### Query the database system catalog to retrieve table metadata

##### You can verify that the table creation was successful by retrieving the list of all tables in your schema and checking whether the SCHOOLS table was created


In [4]:
# type in your query to retrieve list of all tables in the database for your db2 schema (username)
%sql SELECT * FROM SYSCAT.TABLES WHERE TABNAME = 'CHICAGO_SCHOOL_GRADES'

 * ibm_db_sa://kkj13779:***@824dfd4d-99de-440d-9991-629c01b3832d.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:30119/bludb
Done.


[('KKJ13779', 'CHICAGO_SCHOOL_GRADES', 'KKJ13779', 'U', 'T', 'N', None, None, None, None, datetime.datetime(2023, 3, 29, 11, 19, 47, 637249), datetime.datetime(2023, 3, 29, 12, 43, 25, 69749), datetime.datetime(2023, 3, 29, 12, 43, 25, 69758), datetime.datetime(2023, 3, 29, 11, 29, 20, 716213), 79, 14, 2021, 566, 15, 0, 16, -1, -1, -1, 0, 'KKJ13779SPACE1', None, None, 0, 0, 0, 0, 0, 0, 0, 'N', 'YYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY', 1, ' ', '0', -1, 'N', ' ', None, 'R', ' ', 'N', '                                ', None, 'N', ' ', 'F', None, 0, 'N', 999, 0, 0.0, 843, 0.0, None, 1208, 'SYSIBM', 'IDENTITY', 'SYSIBM', 'IDENTITY', ' ', 0, datetime.datetime(2023, 3, 29, 11, 19, 47, 637249), 0, ' ', None, None, 'N', 'KKJ13779', ' ', ' ', ' ', datetime.date(2023, 3, 29), ' ', 'N', 'R', 'N', -1.0, None)]

Double-click **here** for a hint

<!--
In Db2 the system catalog table called SYSCAT.TABLES contains the table metadata
-->


Double-click **here** for the solution.

<!-- Solution:

%sql select TABSCHEMA, TABNAME, CREATE_TIME from SYSCAT.TABLES where TABSCHEMA='YOUR-DB2-USERNAME'

or, you can retrieve list of all tables where the schema name is not one of the system created ones:

%sql select TABSCHEMA, TABNAME, CREATE_TIME from SYSCAT.TABLES \
      where TABSCHEMA not in ('SYSIBM', 'SYSCAT', 'SYSSTAT', 'SYSIBMADM', 'SYSTOOLS', 'SYSPUBLIC')
      
or, just query for a specifc table that you want to verify exists in the database
%sql select * from SYSCAT.TABLES where TABNAME = 'SCHOOLS'

-->


### Query the database system catalog to retrieve column metadata

##### The SCHOOLS table contains a large number of columns. How many columns does this table have?


In [5]:
# type in your query to retrieve the number of columns in the SCHOOLS table

%sql SELECT * FROM SYSCAT.COLUMNS WHERE TABNAME = 'CHICAGO_SCHOOL_GRADES';

 * ibm_db_sa://kkj13779:***@824dfd4d-99de-440d-9991-629c01b3832d.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:30119/bludb
Done.


[('KKJ13779', 'CHICAGO_SCHOOL_GRADES', 'SCHOOL_ID', 0, 'SYSIBM  ', 'INTEGER', 4, 0, None, None, None, 'Y', 0, None, None, ' ', ' ', 566, '610543', '609674', 5, None, 0, 20, 10, 0, None, None, None, None, None, None, None, None, ' ', 0, -1, 'N', 'N', ' ', None, 'O', None, -1.0, -1, -1, -1, None, None, 'N', 'N', 'N', -1, -1.0, None, None, 'N', None),
 ('KKJ13779', 'CHICAGO_SCHOOL_GRADES', 'NAME_OF_SCHOOL', 1, 'SYSIBM  ', 'VARCHAR', 64, 0, 'OCTETS', 64, None, 'Y', 1208, 'SYSIBM', 'IDENTITY', ' ', ' ', 566, "'Woodlawn Community Elementary Sch'", "'Abraham Lincoln Elementary School'", 41, None, 0, 20, 10, 0, None, None, None, None, None, None, None, None, ' ', 0, -1, 'N', 'N', ' ', None, 'O', None, -1.0, -1, -1, 36, None, None, 'N', 'N', 'N', -1, -1.0, None, None, 'N', None),
 ('KKJ13779', 'CHICAGO_SCHOOL_GRADES', 'Elementary, Middle, or High School', 2, 'SYSIBM  ', 'VARCHAR', 2, 0, 'OCTETS', 2, None, 'Y', 1208, 'SYSIBM', 'IDENTITY', ' ', ' ', 3, "'MS'", "'ES'", 7, None, 0, 20, 10, 0, None, None, None, None, None, None, None, None, ' ', 0, -1, 'N', 'N', ' ', None, 'O', None, -1.0, -1, -1, 2, None, None, 'N', 'N', 'N', -1, -1.0, None, None, 'N', None),
 ('KKJ13779', 'CHICAGO_SCHOOL_GRADES', 'STREET_ADDRESS', 3, 'SYSIBM  ', 'VARCHAR', 29, 0, 'OCTETS', 29, None, 'Y', 1208, 'SYSIBM', 'IDENTITY', ' ', ' ', 566, "'9912 S Avenue H'", "'10015 S Leavitt St'", 22, None, 0, 20, 10, 0, None, None, None, None, None, None, None, None, ' ', 0, -1, 'N', 'N', ' ', None, 'O', None, -1.0, -1, -1, 17, None, None, 'N', 'N', 'N', -1, -1.0, None, None, 'N', None),
 ('KKJ13779', 'CHICAGO_SCHOOL_GRADES', 'CITY', 4, 'SYSIBM  ', 'VARCHAR', 7, 0, 'OCTETS', 7, None, 'Y', 1208, 'SYSIBM', 'IDENTITY', ' ', ' ', 1, "'Chicago'", "'Chicago'", 12, None, 0, 20, 10, 0, None, None, None, None, None, None, None, None, ' ', 0, -1, 'N', 'N', ' ', None, 'O', None, -1.0, -1, -1, 7, None, None, 'N', 'N', 'N', -1, -1.0, None, None, 'N', None),
 ('KKJ13779', 'CHICAGO_SCHOOL_GRADES', 'STATE', 5, 'SYSIBM  ', 'VARCHAR', 2, 0, 'OCTETS', 2, None, 'Y', 1208, 'SYSIBM', 'IDENTITY', ' ', ' ', 1, "'IL'", "'IL'", 7, None, 0, 20, 10, 0, None, None, None, None, None, None, None, None, ' ', 0, -1, 'N', 'N', ' ', None, 'O', None, -1.0, -1, -1, 2, None, None, 'N', 'N', 'N', -1, -1.0, None, None, 'N', None),
 ('KKJ13779', 'CHICAGO_SCHOOL_GRADES', 'ZIP_CODE', 6, 'SYSIBM  ', 'INTEGER', 4, 0, None, None, None, 'Y', 0, None, None, ' ', ' ', 51, '60707', '60607', 5, None, 0, 20, 10, 0, None, None, None, None, None, None, None, None, ' ', 0, -1, 'N', 'N', ' ', None, 'O', None, -1.0, -1, -1, -1, None, None, 'N', 'N', 'N', -1, -1.0, None, None, 'N', None),
 ('KKJ13779', 'CHICAGO_SCHOOL_GRADES', 'PHONE_NUMBER', 7, 'SYSIBM  ', 'VARCHAR', 14, 0, 'OCTETS', 14, None, 'Y', 1208, 'SYSIBM', 'IDENTITY', ' ', ' ', 566, "'(773) 535-9930'", "'(773) 534-0146'", 19, None, 0, 20, 10, 0, None, None, None, None, None, None, None, None, ' ', 0, -1, 'N', 'N', ' ', None, 'O', None, -1.0, -1, -1, 14, None, None, 'N', 'N', 'N', -1, -1.0, None, None, 'N', None),
 ('KKJ13779', 'CHICAGO_SCHOOL_GRADES', 'LINK', 8, 'SYSIBM  ', 'VARCHAR', 78, 0, 'OCTETS', 78, None, 'Y', 1208, 'SYSIBM', 'IDENTITY', ' ', ' ', 2, "''", "'http://schoolreports.cps.edu/Scho'", 83, None, 0, 20, 10, 1, None, None, None, None, None, None, None, None, ' ', 0, -1, 'N', 'N', ' ', None, 'O', None, -1.0, -1, -1, 78, None, None, 'N', 'N', 'N', -1, -1.0, None, None, 'N', None),
 ('KKJ13779', 'CHICAGO_SCHOOL_GRADES', 'NETWORK_MANAGER', 9, 'SYSIBM  ', 'VARCHAR', 40, 0, 'OCTETS', 40, None, 'Y', 1208, 'SYSIBM', 'IDENTITY', ' ', ' ', 20, "'Southwest Side High School Networ'", "'Austin-North Lawndale Elementary '", 35, None, 0, 20, 10, 0, None, None, None, None, None, None, None, None, ' ', 0, -1, 'N', 'N', ' ', None, 'O', None, -1.0, -1, -1, 30, None, None, 'N', 'N', 'N', -1, -1.0, None, None, 'N', None),
 ('KKJ13779', 'CHICAGO_SCHOOL_GRADES', 'COLLABORATIVE_NAME', 10, 'SYSIBM  ', 'VARCHAR', 34, 0, 'OCTETS', 34, None, 'Y', 1208, 'SYSIBM', 'IDENTITY', ' ', ' ', 5, "'S

Double-click **here** for a hint

<!--
In Db2 the system catalog table called SYSCAT.COLUMNS contains the column metadata
-->


Double-click **here** for the solution.

<!-- Solution:

%sql select count(*) from SYSCAT.COLUMNS where TABNAME = 'SCHOOLS'

-->


Now retrieve the the list of columns in SCHOOLS table and their column type (datatype) and length.


In [6]:
# type in your query to retrieve all column names in the SCHOOLS table along with their datatypes and length

%sql SELECT colname, typename, length FROM (SELECT * FROM SYSCAT.COLUMNS WHERE TABNAME = 'CHICAGO_SCHOOL_GRADES');

 * ibm_db_sa://kkj13779:***@824dfd4d-99de-440d-9991-629c01b3832d.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:30119/bludb
Done.


[('SCHOOL_ID', 'INTEGER', 4),
 ('NAME_OF_SCHOOL', 'VARCHAR', 64),
 ('Elementary, Middle, or High School', 'VARCHAR', 2),
 ('STREET_ADDRESS', 'VARCHAR', 29),
 ('CITY', 'VARCHAR', 7),
 ('STATE', 'VARCHAR', 2),
 ('ZIP_CODE', 'INTEGER', 4),
 ('PHONE_NUMBER', 'VARCHAR', 14),
 ('LINK', 'VARCHAR', 78),
 ('NETWORK_MANAGER', 'VARCHAR', 40),
 ('COLLABORATIVE_NAME', 'VARCHAR', 34),
 ('ADEQUATE_YEARLY_PROGRESS_MADE_', 'VARCHAR', 3),
 ('TRACK_SCHEDULE', 'VARCHAR', 12),
 ('CPS_PERFORMANCE_POLICY_STATUS', 'VARCHAR', 16),
 ('CPS_PERFORMANCE_POLICY_LEVEL', 'VARCHAR', 15),
 ('HEALTHY_SCHOOL_CERTIFIED', 'VARCHAR', 3),
 ('SAFETY_ICON', 'VARCHAR', 11),
 ('SAFETY_SCORE', 'SMALLINT', 2),
 ('FAMILY_INVOLVEMENT_ICON', 'VARCHAR', 11),
 ('FAMILY_INVOLVEMENT_SCORE', 'VARCHAR', 3),
 ('ENVIRONMENT_ICON', 'VARCHAR', 11),
 ('ENVIRONMENT_SCORE', 'SMALLINT', 2),
 ('INSTRUCTION_ICON', 'VARCHAR', 11),
 ('INSTRUCTION_SCORE', 'SMALLINT', 2),
 ('LEADERS_ICON', 'VARCHAR', 4),
 ('LEADERS_SCORE', 'VARCHAR', 3),
 ('TEACHERS_ICON', 'VARCHAR', 11),
 ('TEACHERS_SCORE', 'VARCHAR', 3),
 ('PARENT_ENGAGEMENT_ICON', 'VARCHAR', 7),
 ('PARENT_ENGAGEMENT_SCORE', 'VARCHAR', 3),
 ('PARENT_ENVIRONMENT_ICON', 'VARCHAR', 7),
 ('PARENT_ENVIRONMENT_SCORE', 'VARCHAR', 3),
 ('AVERAGE_STUDENT_ATTENDANCE', 'VARCHAR', 6),
 ('RATE_OF_MISCONDUCTS__PER_100_STUDENTS_', 'DECIMAL', 5),
 ('AVERAGE_TEACHER_ATTENDANCE', 'VARCHAR', 6),
 ('INDIVIDUALIZED_EDUCATION_PROGRAM_COMPLIANCE_RATE', 'VARCHAR', 7),
 ('PK_2_LITERACY__', 'VARCHAR', 4),
 ('PK_2_MATH__', 'VARCHAR', 4),
 ('GR3_5_GRADE_LEVEL_MATH__', 'VARCHAR', 4),
 ('GR3_5_GRADE_LEVEL_READ__', 'VARCHAR', 4),
 ('GR3_5_KEEP_PACE_READ__', 'VARCHAR', 4),
 ('GR3_5_KEEP_PACE_MATH__', 'VARCHAR', 4),
 ('GR6_8_GRADE_LEVEL_MATH__', 'VARCHAR', 4),
 ('GR6_8_GRADE_LEVEL_READ__', 'VARCHAR', 4),
 ('GR6_8_KEEP_PACE_MATH_', 'VARCHAR', 4),
 ('GR6_8_KEEP_PACE_READ__', 'VARCHAR', 4),
 ('GR_8_EXPLORE_MATH__', 'VARCHAR', 4),
 ('GR_8_EXPLORE_READ__', 'VARCHAR', 4),
 ('ISAT_EXCEEDING_MATH__', 'DECIMAL', 4),
 ('ISAT_EXCEEDING_READING__', 'DECIMAL', 4),
 ('ISAT_VALUE_ADD_MATH', 'DECIMAL', 3),
 ('ISAT_VALUE_ADD_READ', 'DECIMAL', 3),
 ('ISAT_VALUE_ADD_COLOR_MATH', 'VARCHAR', 6),
 ('ISAT_VALUE_ADD_COLOR_READ', 'VARCHAR', 6),
 ('STUDENTS_TAKING__ALGEBRA__', 'VARCHAR', 4),
 ('STUDENTS_PASSING__ALGEBRA__', 'VARCHAR', 4),
 ('9th Grade EXPLORE (2009)', 'VARCHAR', 4),
 ('9th Grade EXPLORE (2010)', 'VARCHAR', 4),
 ('10th Grade PLAN (2009)', 'VARCHAR', 4),
 ('10th Grade PLAN (2010)', 'VARCHAR', 4),
 ('NET_CHANGE_EXPLORE_AND_PLAN', 'VARCHAR', 3),
 ('11th Grade Average ACT (2011)', 'VARCHAR', 4),
 ('NET_CHANGE_PLAN_AND_ACT', 'VARCHAR', 3),
 ('COLLEGE_ELIGIBILITY__', 'VARCHAR', 4),
 ('GRADUATION_RATE__', 'VARCHAR', 4),
 ('COLLEGE_ENROLLMENT_RATE__', 'VARCHAR', 4),
 ('COLLEGE_ENROLLMENT', 'SMALLINT', 2),
 ('GENERAL_SERVICES_ROUTE', 'SMALLINT', 2),
 ('FRESHMAN_ON_TRACK_RATE__', 'VARCHAR', 4),
 ('X_COORDINATE', 'DECIMAL', 13),
 ('Y_COORDINATE', 'DECIMAL', 13),
 ('LATITUDE', 'DECIMAL', 18),
 ('LONGITUDE', 'DECIMAL', 18),
 ('COMMUNITY_AREA_NUMBER', 'SMALLINT', 2),
 ('COMMUNITY_AREA_NAME', 'VARCHAR', 22),
 ('WARD', 'SMALLINT', 2),
 ('POLICE_DISTRICT', 'SMALLINT', 2),
 ('LOCATION', 'VARCHAR', 27),
 ('SUM_COLLEGE_ENROLLMENT', 'BIGINT', 8)]

In [7]:
%sql SELECT * FROM CHICAGO_SCHOOL_GRADES LIMIT 10;

 * ibm_db_sa://kkj13779:***@824dfd4d-99de-440d-9991-629c01b3832d.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:30119/bludb
Done.


[(610038, 'Abraham Lincoln Elementary School', 'ES', '615 W Kemper Pl', 'Chicago', 'IL', 60614, '(773) 534-5720', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610038.pdf', 'Fullerton Elementary Network', 'NORTH-NORTHWEST SIDE COLLABORATIVE', 'No', 'Standard', 'Not on Probation', 'Level 1', 'Yes', 'Very Strong', 99, 'Very Strong', '99', 'Strong', 74, 'Strong', 66, 'Weak', '65', 'Strong', '70', 'Strong', '56', 'Average', '47', '96.00%', Decimal('2.0'), '96.40%', '95.80%', '80.1', '43.3', '89.6', '84.9', '60.7', '62.6', '81.9', '85.2', '52', '62.4', '66.3', '77.9', Decimal('69.7'), Decimal('64.4'), Decimal('0.2'), Decimal('0.9'), 'Yellow', 'Green', '67.1', '54.5', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 813, 33, 'NDA', Decimal('1171699.458'), Decimal('1915829.428'), Decimal('41.92449696'), Decimal('-87.64452163'), 7, 'LINCOLN PARK', 43, 18, '(41.92449696, -87.64452163)', None),
 (610281, 'Adam Clayton Powell Paideia Community Academy Elementary School', 'ES', '7511 S South Shore Dr', 'Chicago', 'IL', 60649, '(773) 535-6650', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610281.pdf', 'Skyway Elementary Network', 'SOUTH SIDE COLLABORATIVE', 'No', 'Track_E', 'Not on Probation', 'Level 1', 'No', 'Average', 54, 'Strong', '66', 'Strong', 74, 'Very Strong', 84, 'Weak', '63', 'Strong', '76', 'Weak', '46', 'Average', '50', '95.60%', Decimal('15.7'), '95.30%', '100.00%', '62.4', '51.7', '21.9', '15.1', '29', '42.8', '38.5', '27.4', '44.8', '42.7', '14.1', '34.4', Decimal('16.8'), Decimal('16.5'), Decimal('0.7'), Decimal('1.4'), 'Green', 'Green', '17.2', '27.3', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 521, 46, 'NDA', Decimal('1196129.985'), Decimal('1856209.466'), Decimal('41.76032435'), Decimal('-87.55673627'), 43, 'SOUTH SHORE', 7, 4, '(41.76032435, -87.55673627)', None),
 (610185, 'Adlai E Stevenson Elementary School', 'ES', '8010 S Kostner Ave', 'Chicago', 'IL', 60652, '(773) 535-2280', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610185.pdf', 'Midway Elementary Network', 'SOUTHWEST SIDE COLLABORATIVE', 'No', 'Standard', 'Not on Probation', 'Level 2', 'No', 'Strong', 61, 'NDA', 'NDA', 'Average', 50, 'Weak', 36, 'Weak', 'NDA', 'NDA', 'NDA', 'Average', '47', 'Weak', '41', '95.70%', Decimal('2.3'), '94.70%', '98.30%', '53.7', '26.6', '38.3', '34.7', '43.7', '57.3', '48.8', '39.2', '46.8', '44', '7.5', '21.9', Decimal('18.3'), Decimal('15.5'), Decimal('-0.9'), Decimal('-1.0'), 'Red', 'Red', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 1324, 44, 'NDA', Decimal('1148427.165'), Decimal('1851012.215'), Decimal('41.74711093'), Decimal('-87.73170248'), 70, 'ASHBURN', 13, 8, '(41.74711093, -87.73170248)', None),
 (609993, 'Agustin Lara Elementary Academy', 'ES', '4619 S Wolcott Ave', 'Chicago', 'IL', 60609, '(773) 535-4389', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_609993.pdf', 'Pershing Elementary Network', 'SOUTHWEST SIDE COLLABORATIVE', 'No', 'Track_E', 'Not on Probation', 'Level 1', 'No', 'Average', 56, 'Average', '44', 'Average', 45, 'Weak', 37, 'Weak', '65', 'Average', '48', 'Average', '53', 'Strong', '58', '95.50%', Decimal('10.4'), '95.80%', '100.00%', '76.9', 'NDA', '26', '24.7', '61.8', '49.7', '39.2', '27.2', '69.7', '60.6', '9.1', '18.2', Decimal('11.1'), Decimal('9.6'), Decimal('0.9'), Decimal('2.4'), 'Green', 'Green', '42.9', '25', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 556, 42, 'NDA', Decimal('1164504.290'), Decimal('1873959.199'), Decimal('41.80975690'), Decimal('-87.67214460'), 61, 'NEW CITY', 20, 9, '(41.8097569, -87.6721446)', None),
 (610513, 'Air Force Academy High School', 'HS', '3630 S Wells St', 'Chicago', 'IL', 60609, '(773) 535-1590', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610513.pdf', 'Southwest Side High School Network', 'SOUTHWEST SIDE COLLABORATIVE', 'NDA', 'Stand

Double-click **here** for the solution.

<!-- Solution:

%sql select COLNAME, TYPENAME, LENGTH from SYSCAT.COLUMNS where TABNAME = 'SCHOOLS'

or

%sql select distinct(NAME), COLTYPE, LENGTH from SYSIBM.SYSCOLUMNS where TBNAME = 'SCHOOLS'

-->


### Questions

1.  Is the column name for the "SCHOOL ID" attribute in upper or mixed case?
2.  What is the name of "Community Area Name" column in your table? Does it have spaces?
3.  Are there any columns in whose names the spaces and paranthesis (round brackets) have been replaced by the underscore character "_"?


## Problems

### Problem 1

##### How many Elementary Schools are in the dataset?


In [8]:
%sql SELECT COUNT(SCHOOL_ID) FROM CHICAGO_SCHOOL_GRADES WHERE "Elementary, Middle, or High School" = 'ES';

 * ibm_db_sa://kkj13779:***@824dfd4d-99de-440d-9991-629c01b3832d.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:30119/bludb
Done.


[(461,)]

Double-click **here** for a hint

<!--
Which column specifies the school type e.g. 'ES', 'MS', 'HS'? ("Elementary School, Middle School, High School")
-->


Double-click **here** for another hint

<!--
Does the column name have mixed case, spaces or other special characters?
If so, ensure you use double quotes around the "Name of the Column"
-->


Double-click **here** for the solution.

<!-- Solution:

%sql select count(*) from SCHOOLS where "Elementary, Middle, or High School" = 'ES'

Correct answer: 462

-->


### Problem 2

##### What is the highest Safety Score?


In [9]:
%sql SELECT MAX(safety_score) FROM CHICAGO_SCHOOL_GRADES;

 * ibm_db_sa://kkj13779:***@824dfd4d-99de-440d-9991-629c01b3832d.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:30119/bludb
Done.


[(99,)]

Double-click **here** for a hint

<!--
Use the MAX() function
-->


Double-click **here** for the solution.

<!-- Hint:

%sql select MAX(Safety_Score) AS MAX_SAFETY_SCORE from SCHOOLS

Correct answer: 99
-->


### Problem 3

##### Which schools have highest Safety Score?


In [10]:
%sql SELECT school_id, name_of_school FROM CHICAGO_SCHOOL_GRADES WHERE safety_score = 99;

 * ibm_db_sa://kkj13779:***@824dfd4d-99de-440d-9991-629c01b3832d.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:30119/bludb
Done.


[(610038, 'Abraham Lincoln Elementary School'),
 (609799, 'Alexander Graham Bell Elementary School'),
 (610084, 'Annie Keller Elementary Gifted Magnet School'),
 (609820, 'Augustus H Burley Elementary School'),
 (610132, 'Edgar Allan Poe Elementary Classical School'),
 (609901, 'Edgebrook Elementary School'),
 (610073, 'Ellen Mitchell Elementary School'),
 (610066, 'James E McDade Elementary Classical School'),
 (609803, 'James G Blaine Elementary School'),
 (610033, 'LaSalle Elementary Language Academy'),
 (610355, 'Mary E Courtenay Elementary Language Arts Center'),
 (609749, 'Northside College Preparatory High School'),
 (609744, 'Northside Learning Center High School'),
 (610099, 'Norwood Park Elementary School'),
 (610105, 'Oriole Park Elementary School'),
 (610155, 'Sauganash Elementary School'),
 (609880, 'Stephen Decatur Classical Elementary School'),
 (610249, 'Talman Elementary School'),
 (610230, 'Wildwood Elementary School')]

In [11]:
%sql DELETE FROM CHICAGO_SCHOOL_GRADES WHERE average_student_attendance IS NULL;

 * ibm_db_sa://kkj13779:***@824dfd4d-99de-440d-9991-629c01b3832d.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:30119/bludb
0 rows affected.


[]

Double-click **here** for the solution.

<!-- Solution:
In the previous problem we found out that the highest Safety Score is 99, so we can use that as an input in the where clause:

%sql select Name_of_School, Safety_Score from SCHOOLS where Safety_Score = 99

or, a better way:

%sql select Name_of_School, Safety_Score from SCHOOLS where \
  Safety_Score= (select MAX(Safety_Score) from SCHOOLS)


Correct answer: several schools with with Safety Score of 99.
-->


### Problem 4

##### What are the top 10 schools with the highest "Average Student Attendance"?


In [12]:
%sql SELECT school_id, name_of_school, average_student_attendance FROM CHICAGO_SCHOOL_GRADES ORDER BY average_student_attendance DESC LIMIT 10;

 * ibm_db_sa://kkj13779:***@824dfd4d-99de-440d-9991-629c01b3832d.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:30119/bludb
Done.


[(609959, 'John Charles Haines Elementary School', '98.40%'),
 (610217, 'James Ward Elementary School', '97.80%'),
 (610132, 'Edgar Allan Poe Elementary Classical School', '97.60%'),
 (609842, 'Rachel Carson Elementary School', '97.60%'),
 (610329, 'Orozco Fine Arts & Sciences Elementary School', '97.60%'),
 (610084, 'Annie Keller Elementary Gifted Magnet School', '97.50%'),
 (610060, 'Andrew Jackson Elementary Language Academy', '97.40%'),
 (610298, 'Lenart Elementary Regional Gifted Center', '97.40%'),
 (610515, 'Disney II Magnet School', '97.30%'),
 (610207, 'John H Vanderpoel Elementary Magnet School', '97.20%')]

Double-click **here** for the solution.

<!-- Solution:

%sql select Name_of_School, Average_Student_Attendance from SCHOOLS \
    order by Average_Student_Attendance desc nulls last limit 10 

-->


### Problem 5

##### Retrieve the list of 5 Schools with the lowest Average Student Attendance sorted in ascending order based on attendance


In [13]:
%sql SELECT school_id, name_of_school, average_student_attendance FROM chicago_school_grades ORDER BY average_student_attendance \
        FETCH FIRST 5 ROWS ONLY


 * ibm_db_sa://kkj13779:***@824dfd4d-99de-440d-9991-629c01b3832d.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:30119/bludb
Done.


[(609702, 'Richard T Crane Technical Preparatory High School', '57.90%'),
 (609871, 'Barbara Vick Early Childhood & Family Center', '60.90%'),
 (609736, 'Dyett High School', '62.50%'),
 (609727, 'Wendell Phillips Academy High School', '63.00%'),
 (610389, 'Orr Academy High School', '66.30%')]

Double-click **here** for the solution.

<!-- Solution:

%sql SELECT Name_of_School, Average_Student_Attendance  \
     from SCHOOLS \
     order by Average_Student_Attendance \
     fetch first 5 rows only

-->


### Problem 6

##### Now remove the '%' sign from the above result set for Average Student Attendance column


In [14]:
%sql select school_id, name_of_school, replace(average_student_attendance, '%', '') from chicago_school_grades order by average_student_attendance \
    fetch first 5 rows only;

 * ibm_db_sa://kkj13779:***@824dfd4d-99de-440d-9991-629c01b3832d.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:30119/bludb
Done.


[(609702, 'Richard T Crane Technical Preparatory High School', '57.90'),
 (609871, 'Barbara Vick Early Childhood & Family Center', '60.90'),
 (609736, 'Dyett High School', '62.50'),
 (609727, 'Wendell Phillips Academy High School', '63.00'),
 (610389, 'Orr Academy High School', '66.30')]

Double-click **here** for a hint

<!--
Use the REPLACE() function to replace '%' with ''
See documentation for this function at:
https://www.ibm.com/support/knowledgecenter/en/SSEPGG_10.5.0/com.ibm.db2.luw.sql.ref.doc/doc/r0000843.html
-->


Double-click **here** for the solution.

<!-- Hint:

%sql SELECT Name_of_School, REPLACE(Average_Student_Attendance, '%', '') \
     from SCHOOLS \
     order by Average_Student_Attendance \
     fetch first 5 rows only

-->


### Problem 7

##### Which Schools have Average Student Attendance lower than 70%?


In [15]:
%sql select school_id, name_of_school, average_student_attendance from chicago_school_grades where average_student_attendance < '70%' \
    order by average_student_attendance desc;

 * ibm_db_sa://kkj13779:***@824dfd4d-99de-440d-9991-629c01b3832d.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:30119/bludb
Done.


[(609759, 'Roberto Clemente Community Academy High School', '69.60%'),
 (609674, 'Chicago Vocational Career Academy High School', '68.80%'),
 (609722, 'Manley Career Academy High School', '66.80%'),
 (610389, 'Orr Academy High School', '66.30%'),
 (609727, 'Wendell Phillips Academy High School', '63.00%'),
 (609736, 'Dyett High School', '62.50%'),
 (609871, 'Barbara Vick Early Childhood & Family Center', '60.90%'),
 (609702, 'Richard T Crane Technical Preparatory High School', '57.90%')]

Double-click **here** for a hint

<!--
The datatype of the "Average_Student_Attendance" column is varchar.
So you cannot use it as is in the where clause for a numeric comparison.
First use the CAST() function to cast it as a DECIMAL or DOUBLE
e.g. CAST("Column_Name" as DOUBLE)
or simply: DECIMAL("Column_Name")
-->


Double-click **here** for another hint

<!--
Don't forget the '%' age sign needs to be removed before casting
-->


Double-click **here** for the solution.

<!-- Solution:

%sql SELECT Name_of_School, Average_Student_Attendance  \
     from SCHOOLS \
     where CAST ( REPLACE(Average_Student_Attendance, '%', '') AS DOUBLE ) < 70 \
     order by Average_Student_Attendance
     
or,

%sql SELECT Name_of_School, Average_Student_Attendance  \
     from SCHOOLS \
     where DECIMAL ( REPLACE(Average_Student_Attendance, '%', '') ) < 70 \
     order by Average_Student_Attendance

-->


### Problem 8

##### Get the total College Enrollment for each Community Area


In [17]:
import pandas as pd
%sql select community_area_name, SUM(college_enrollment) as sum_college_enrollment from chicago_school_grades group by community_area_name;

 * ibm_db_sa://kkj13779:***@824dfd4d-99de-440d-9991-629c01b3832d.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:30119/bludb
Done.


[('ALBANY PARK', 6864),
 ('ARCHER HEIGHTS', 4823),
 ('ARMOUR SQUARE', 1458),
 ('ASHBURN', 6483),
 ('AUBURN GRESHAM', 4175),
 ('AUSTIN', 10933),
 ('AVALON PARK', 1522),
 ('AVONDALE', 3640),
 ('BELMONT CRAGIN', 14386),
 ('BEVERLY', 1636),
 ('BRIDGEPORT', 3167),
 ('BRIGHTON PARK', 9647),
 ('BURNSIDE', 549),
 ('CALUMET HEIGHTS', 1568),
 ('CHATHAM', 5042),
 ('CHICAGO LAWN', 7086),
 ('CLEARING', 2085),
 ('DOUGLAS', 4670),
 ('DUNNING', 4568),
 ('EAST GARFIELD PARK', 5337),
 ('EAST SIDE', 5305),
 ('EDGEWATER', 4600),
 ('EDISON PARK', 910),
 ('ENGLEWOOD', 6832),
 ('FOREST GLEN', 1431),
 ('FULLER PARK', 531),
 ('GAGE PARK', 9915),
 ('GARFIELD RIDGE', 4552),
 ('GRAND BOULEVARD', 2809),
 ('GREATER GRAND CROSSING', 4051),
 ('HEGEWISCH', 963),
 ('HERMOSA', 3975),
 ('HUMBOLDT PARK', 8620),
 ('HYDE PARK', 1930),
 ('IRVING PARK', 7764),
 ('JEFFERSON PARK', 1755),
 ('KENWOOD', 4287),
 ('LAKE VIEW', 7055),
 ('LINCOLN PARK', 5615),
 ('LINCOLN SQUARE', 4132),
 ('LOGAN SQUARE', 7351),
 ('LOOP', 871),
 ('LOWER WEST SIDE', 7257),
 ('MCKINLEY PARK', 1341),
 ('MONTCLARE', 1317),
 ('MORGAN PARK', 3271),
 ('MOUNT GREENWOOD', 2091),
 ('NEAR NORTH SIDE', 3362),
 ('NEAR SOUTH SIDE', 1378),
 ('NEAR WEST SIDE', 7975),
 ('NEW CITY', 7922),
 ('NORTH CENTER', 7541),
 ('NORTH LAWNDALE', 5146),
 ('NORTH PARK', 4210),
 ('NORWOOD PARK', 6469),
 ('OAKLAND', 140),
 ('OHARE', 786),
 ('PORTAGE PARK', 6954),
 ('PULLMAN', 1620),
 ('RIVERDALE', 1547),
 ('ROGERS PARK', 4068),
 ('ROSELAND', 7020),
 ('SOUTH CHICAGO', 4043),
 ('SOUTH DEERING', 1859),
 ('SOUTH LAWNDALE', 14793),
 ('SOUTH SHORE', 4543),
 ('UPTOWN', 4388),
 ('WASHINGTON HEIGHTS', 4006),
 ('WASHINGTON PARK', 2648),
 ('WEST ELSDON', 3700),
 ('WEST ENGLEWOOD', 5946),
 ('WEST GARFIELD PARK', 2622),
 ('WEST LAWN', 4207),
 ('WEST PULLMAN', 3240),
 ('WEST RIDGE', 8197),
 ('WEST TOWN', 9429),
 ('WOODLAWN', 4206)]

Double-click **here** for a hint

<!--
Verify the exact name of the Enrollment column in the database
Use the SUM() function to add up the Enrollments for each Community Area
-->


Double-click **here** for another hint

<!--
Don't forget to group by the Community Area
-->


Double-click **here** for the solution.

<!-- Solution:

%sql select Community_Area_Name, sum(College_Enrollment) AS TOTAL_ENROLLMENT \
   from SCHOOLS \
   group by Community_Area_Name 

-->


### Problem 9

##### Get the 5 Community Areas with the least total College Enrollment  sorted in ascending order


In [18]:
%sql select * from (select community_area_name, SUM(college_enrollment) as sum_college_enrollment from chicago_school_grades group by community_area_name) order by sum_college_enrollment fetch first 5 rows only;

 * ibm_db_sa://kkj13779:***@824dfd4d-99de-440d-9991-629c01b3832d.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:30119/bludb
Done.


[('OAKLAND', 140),
 ('FULLER PARK', 531),
 ('BURNSIDE', 549),
 ('OHARE', 786),
 ('LOOP', 871)]

Double-click **here** for a hint

<!--
Order the previous query and limit the number of rows you fetch
-->


Double-click **here** for the solution.

<!-- Solution:

%sql select Community_Area_Name, sum(College_Enrollment) AS TOTAL_ENROLLMENT \
   from SCHOOLS \
   group by Community_Area_Name \
   order by TOTAL_ENROLLMENT asc \
   fetch first 5 rows only

-->


### Problem 10

##### List 5 schools with lowest safety score.


In [19]:
%sql select school_id, name_of_school, safety_score from chicago_school_grades order by safety_score fetch first 5 rows only;

 * ibm_db_sa://kkj13779:***@824dfd4d-99de-440d-9991-629c01b3832d.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:30119/bludb
Done.


[(609819, 'Edmond Burke Elementary School', 1),
 (610108, "Luke O'Toole Elementary School", 5),
 (610202, 'George W Tilton Elementary School', 6),
 (609927, 'Foster Park Elementary School', 11),
 (609712, 'Emil G Hirsch Metropolitan High School', 13)]

Double-click **here** for the solution.

<!-- Solution:

%sql SELECT name_of_school, safety_score \
FROM schools \
ORDER BY safety_score \
LIMIT 5
-->


### Problem 11

##### Get the hardship index for the community area which has College Enrollment of 4368


In [20]:
%%sql 
select hardship_index 
   from chicago_socioeconomic_data CD, chicago_school_grades CPS 
   where CD.ca = CPS.community_area_number 
      and college_enrollment = 4368

 * ibm_db_sa://kkj13779:***@824dfd4d-99de-440d-9991-629c01b3832d.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:30119/bludb
Done.


[(6.0,)]

Double-click **here** for the solution.

<!-- Solution:
NOTE: For this solution to work the CHICAGO_SOCIOECONOMIC_DATA table 
      as created in the last lab of Week 3 should already exist

%%sql 
select hardship_index 
   from chicago_socioeconomic_data CD, schools CPS 
   where CD.ca = CPS.community_area_number 
      and college_enrollment = 4368

-->


### Problem 12

##### Get the hardship index for the community area which has the school with the  highest enrollment.


In [21]:
%%sql
select hardship_index, community_area_name from chicago_socioeconomic_data se, chicago_school_grades sc
where se.community_area_name = sc.community_area_name and sc.college_enrollment = MAX(sc.college_enrollment);

 * ibm_db_sa://kkj13779:***@824dfd4d-99de-440d-9991-629c01b3832d.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:30119/bludb
(ibm_db_dbi.ProgrammingError) ibm_db_dbi::ProgrammingError: SQLNumResultCols failed: [IBM][CLI Driver][DB2/LINUXX8664] SQL0120N  Invalid use of an aggregate function or OLAP function.  SQLSTATE=42903 SQLCODE=-120
[SQL: select hardship_index, community_area_name from chicago_socioeconomic_data se, chicago_school_grades sc
where se.community_area_name = sc.community_area_name and sc.college_enrollment = MAX(sc.college_enrollment);]
(Background on this error at: http://sqlalche.me/e/13/f405)


Double-click **here** for the solution.

<!-- Solution:
NOTE: For this solution to work the CHICAGO_SOCIOECONOMIC_DATA table 
      as created in the last lab of Week 3 should already exist

%sql select ca, community_area_name, hardship_index from chicago_socioeconomic_data \
   where ca in \
   ( select community_area_number from schools order by college_enrollment desc limit 1 )

-->


## Summary

##### In this lab you learned how to work with a real word dataset using SQL and Python. You learned how to query columns with spaces or special characters in their names and with mixed case names. You also used built in database functions and practiced how to sort, limit, and order result sets, as well as used sub-queries and worked with multiple tables.


## Author

<a href="https://www.linkedin.com/in/ravahuja/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork22-2022-01-01" target="_blank">Rav Ahuja</a>

## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description                 |
| ----------------- | ------- | ---------- | ---------------------------------- |
| 2021-07-09        | 2.4     | Malika     | Updated connection string  |
| 2021-05-19        | 2.3     | Lakshmi Holla |Updated question     |
| 2021-04-20        | 2.2     | Malika     | Added the libraries                |
| 2020-11-27        | 2.1     | Sannareddy Ramesh    | Modified data sets and added new problems |
| 2020-08-28        | 2.0     | Lavanya    | Moved lab to course repo in GitLab |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
